src/_pytest/python.py

```diff
@@ -151,15 +152,16 @@ def pytest_configure(config):
 @hookimpl(trylast=True)
 def pytest_pyfunc_call(pyfuncitem):
     testfunction = pyfuncitem.obj
-    iscoroutinefunction = getattr(inspect, "iscoroutinefunction", None)
-    if iscoroutinefunction is not None and iscoroutinefunction(testfunction):
-        msg = "Coroutine functions are not natively supported and have been skipped.\n"
+    if iscoroutinefunction(testfunction) or (
+        sys.version_info >= (3, 6) and inspect.isasyncgenfunction(testfunction)
+    ):
+        msg = "async def functions are not natively supported and have been skipped.\n"
         msg += "You need to install a suitable plugin for your async framework, for example:\n"
         msg += "  - pytest-asyncio\n"
         msg += "  - pytest-trio\n"
         msg += "  - pytest-tornasync"
         warnings.warn(PytestUnhandledCoroutineWarning(msg.format(pyfuncitem.nodeid)))
-        skip(msg="coroutine function and no async plugin installed (see warnings)")
+        skip(msg="async def function and no async plugin installed (see warnings)")
     funcargs = pyfuncitem.funcargs
     testargs = {arg: funcargs[arg] for arg in pyfuncitem._fixtureinfo.argnames}
     testfunction(**testargs)
```

Comment:
@bluetech it might also be easier to just 
```suggestion
    assert testfunction(**testargs) is None
```

This would catch generator, async def, async gen and t.i.d.inlineCallbacks all in one go

Reply:
https://github.com/pytest-dev/pytest/issues/5735

Reply:
I don't know enough to have an opinion on this.

If the test function should absolutely never return anything once it reaches here, maybe a warning is appropriate.

Reply:
there are legitimate test functions that return a value as a building block, so this one would be a breaking change

Reply:
> legitimate test functions

Sure I was thinking this would be a deprecation warning rather than a skip

Reply:
Let's leave this alone, given that it is a valid construct I wouldn't even issue a warning.

Reply:
@RonnyPfannschmidt do you have an example?

Reply:
@nicoddemus i would like tests to require a return type annotation if they are used as building blocks to communicate the intent

Reply:
@graingert not off-hand, the test i recall is in a internal repo at a previous employer

Reply:
I think we can make the same call as was made when deprecating direct calls of fixtures and require code to look like:

```
def _prepare_for_test():
    assert something()
    return FooBar()

def test_a()
    _prepare_for_test()

def test_b():
    fb = _prepare_for_test()
    assert fb.something()
```

Reply:
Not sure why you'd need test_a in this case though

Reply:
this is pretty much just code-reuse - test_a would  simply be a "dependency/fixture" for the b test, but ots own a test anyway

Reply:
It's not code-reuse - it's just redundant, there's no need for `test_a`

```
def test_b():
    assert something()
    fb =  FooBar()
    assert fb.something()
```

Reply:
so what, it happens in practice, is used in practice and should have a transition period for users to adapt better structural practices